src/manifold/src/smoothing.cpp

```diff
@@ -501,6 +501,269 @@ class Partition {
 
 namespace manifold {
 
+// sharpenedEdges are referenced to the input Mesh, but the triangles have
+// been sorted in creating the Manifold, so the indices are converted using
+// meshRelation_.
+std::vector<Smoothness> Manifold::Impl::UpdateSharpenedEdges(
+    const std::vector<Smoothness>& sharpenedEdges) const {
+  std::unordered_map<int, int> oldHalfedge2New;
+  for (int tri = 0; tri < NumTri(); ++tri) {
+    int oldTri = meshRelation_.triRef[tri].tri;
+    for (int i : {0, 1, 2}) oldHalfedge2New[3 * oldTri + i] = 3 * tri + i;
+  }
+  std::vector<Smoothness> newSharp = sharpenedEdges;
+  for (Smoothness& edge : newSharp) {
+    edge.halfedge = oldHalfedge2New[edge.halfedge];
+  }
+  return newSharp;
+}
+
+// Find faces containing at least 3 triangles - these will not have
+// interpolated normals - all their vert normals must match their face normal.
+Vec<bool> Manifold::Impl::FlatFaces() const {
+  const int numTri = NumTri();
+  Vec<bool> triIsFlatFace(numTri, false);
+  for (int tri = 0; tri < numTri; ++tri) {
+    const TriRef& ref = meshRelation_.triRef[tri];
+    int faceNeighbors = 0;
+    glm::ivec3 faceTris = {-1, -1, -1};
+    for (const int j : {0, 1, 2}) {
+      const int neighborTri =
+          halfedge_[halfedge_[3 * tri + j].pairedHalfedge].face;
+      const TriRef& jRef = meshRelation_.triRef[neighborTri];
+      if (jRef.SameFace(ref)) {
+        ++faceNeighbors;
+        faceTris[j] = neighborTri;
+      }
+    }
+    if (faceNeighbors > 1) {
+      triIsFlatFace[tri] = true;
```

Comment:
can parallelize this as well?
if we want to avoid repeated work, we should check `faceTris[tri]` before the start of the `j` loop.

Reply:
I think for correctness we can't do the early-out.